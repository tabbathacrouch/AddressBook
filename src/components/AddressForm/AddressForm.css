.form {
  margin-top: 1.5rem;
  margin-left: 1.5rem;
  margin-right: 2.5rem;
  display: flex;
  flex-direction: column;
  font-size: 1rem;
}

.error {
  color: red;
  margin-left: 1rem;
}

.field {
  width: 100%;
  margin: 0.5rem;
  padding: 0.5rem;
  border-style: solid;
  border-width: 1px;
  border-radius: 0.25rem;
  box-shadow: 2px 2px 4px 2px rgba( 0, 0, 0, 0.2);
  font-size: 1rem;
}

.field:hover {
  background-color: #F4F4F4;
}

.city {
  width: 55%;
  margin-right: 2em;
}

.stateName {
  width: 15%;
}

.zip {
  margin-left: 2em;
}

.fieldRow {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.btn-row {
  width: 100%;
  display: flex;
  justify-content: center;
}

.submit-btn {
  border-style: solid;
  border-width: 1px;
  border-radius: 0.25rem;
  box-shadow: 2px 2px 4px 2px rgba( 0, 0, 0, 0.2);
  width: 25%;
  padding: 0.5rem;
  margin: 0.5rem;
  font-size: 1rem;
}

.submit-btn:hover {
  background-color: #333333;
  color: white;
}

.search-form {
  display: flex;
  justify-content: center;
}