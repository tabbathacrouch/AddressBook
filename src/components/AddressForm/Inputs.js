import React from "react";
import { Field, ErrorMessage } from "formik";

export const NameInput = () => (
  <div>
    <Field
      data-testid="name"
      name="name"
      type="text"
      placeholder="Name"
      className="field"
    />
    <ErrorMessage name="name">
      {(errorMessage) => (
        <div className="error" data-testid="nameError">
          {errorMessage}
        </div>
      )}
    </ErrorMessage>
  </div>
);

export const StreetInput = () => (
  <div>
    <Field
      data-testid="street"
      name="street"
      type="text"
      placeholder="Street"
      className="field"
    />
    <ErrorMessage name="street">
      {(errorMessage) => (
        <div className="error" data-testid="streetError">
          {errorMessage}
        </div>
      )}
    </ErrorMessage>
  </div>
);

export const CityInput = () => (
  <div className="city">
    <Field data-testid="city" name="city" type="text" placeholder="City" className="field" />
    <ErrorMessage name="city">
      {(errorMessage) => <div className="error" data-testid="cityError">{errorMessage}</div>}
    </ErrorMessage>
  </div>
);

export const StateInput = () => (
  <div className="stateName">
    <Field data-testid="stateName" name="stateName" type="text" placeholder="State" className="field" />
    <ErrorMessage name="stateName">
      {(errorMessage) => <div className="error" data-testid="stateNameError">{errorMessage}</div>}
    </ErrorMessage>
  </div>
);

export const ZipInput = () => (
  <div className="zip">
    <Field data-testid="zip" name="zip" type="text" placeholder="Zip" className="field" />
    <ErrorMessage name="zip">
      {(errorMessage) => <div className="error" data-testid="zipError">{errorMessage}</div>}
    </ErrorMessage>
  </div>
);
