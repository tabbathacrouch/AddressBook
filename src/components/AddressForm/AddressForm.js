import React from "react";
import { Formik, Form } from "formik";
import * as Yup from "yup";
import "./AddressForm.css";
import {
  NameInput,
  StreetInput,
  CityInput,
  StateInput,
  ZipInput,
} from "./Inputs";


export const addressFormValidator = Yup.object({
  name: Yup.string()
    .max(30, "Must be 30 characters or less")
    .matches(
      /^[aA-zZ\s]+$/,
      "Numbers and special characters are not allowed"
    )
    .required("Required"),
  street: Yup.string()
    .max(40, "Must be 40 characters or less")
    .required("Required"),
  city: Yup.string()
    .max(50, "Must be 50 characters or less")
    .matches(
      /^[aA-zZ\s]+$/,
      "Numbers and special characters are not allowed"
    )
    .required("Required"),
  stateName: Yup.string()
    .max(15, "Must be 15 characters or less")
    .matches(/^[aA-zZ\s]+$/, "Invalid State name")
    .required("Required"),
  zip: Yup.string()
    .matches(
      /^(?=(\D*\d){5}\D*$)(?=[^ ]* ?[^ ]*$)[\d ]*$/,
      "Must contain 5 digits"
    )
    .required("Required"),
});

const AddressForm = ({ addContact }) => {
  return (
    <Formik
      initialValues={{
        id: 0,
        name: "",
        street: "",
        city: "",
        stateName: "",
        zip: "",
      }}
      validationSchema={addressFormValidator}
      onSubmit={(contact, { resetForm }) => {
        addContact(contact);
        const prev = contact.id;
        resetForm({
          values: {
            id: prev + 1,
            name: "",
            street: "",
            city: "",
            stateName: "",
            zip: "",
          },
        });
      }}
    >
      <Form>
        <div className="form">
          <NameInput />
          <StreetInput />
          <div className="fieldRow">
            <CityInput />
            <StateInput />
            <ZipInput />
          </div>
          <div className="btn-row">
            <button type="submit" className="submit-btn">
              Submit
            </button>
          </div>
        </div>
      </Form>
    </Formik>
  );
};

export default AddressForm;
