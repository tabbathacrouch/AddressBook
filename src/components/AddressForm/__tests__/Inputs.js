import React from "react";
import renderer from "react-test-renderer";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { Formik } from "formik";
import { addressFormValidator } from "../AddressForm";
import {
  NameInput,
  StreetInput,
  CityInput,
  StateInput,
  ZipInput,
} from "../Inputs";

const renderWithFormik = (Component) =>
  render(
    <Formik
      validationSchema={addressFormValidator}
      initialValues={{
        name: "",
        street: "",
        city: "",
        stateName: "",
        zip: "",
      }}
    >
      <Component />
    </Formik>
  );

describe("<NameInput />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <Formik validationSchema={addressFormValidator}>
        <NameInput />
      </Formik>
    );
    expect(tree).toMatchSnapshot();
  });

  describe("errors", () => {
    it("should show required error", async () => {
      const { getByTestId } = renderWithFormik(NameInput);

      fireEvent.focus(getByTestId("name"));
      fireEvent.blur(getByTestId("name"));

      await waitFor(() =>
        expect(getByTestId("nameError")).toHaveTextContent("Required")
      );
    });

    it("shows special character error", async () => {
      const { getByTestId } = renderWithFormik(NameInput);

      fireEvent.change(getByTestId("name"), { target: { value: "name@" } });
      fireEvent.blur(getByTestId("name"));

      await waitFor(() =>
        expect(getByTestId("nameError")).toHaveTextContent(
          "Numbers and special characters are not allowed"
        )
      );
    });

    it("shows maxlength error", async () => {
      const { getByTestId } = renderWithFormik(NameInput);
      fireEvent.change(getByTestId("name"), {
        target: { value: "name".repeat(8) },
      });
      fireEvent.blur(getByTestId("name"));

      await waitFor(() =>
        expect(getByTestId("nameError")).toHaveTextContent(
          "Must be 30 characters or less"
        )
      );
    });
  });
});

describe("<StreetInput />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <Formik validationSchema={addressFormValidator}>
        <StreetInput />
      </Formik>
    );
    expect(tree).toMatchSnapshot();
  });

  describe("errors", () => {
    it("should show required error", async () => {
      const { getByTestId } = renderWithFormik(StreetInput);
      fireEvent.focus(getByTestId("street"));
      fireEvent.blur(getByTestId("street"));

      await waitFor(() =>
        expect(getByTestId("streetError")).toHaveTextContent("Required")
      );
    });

    it("shows maxlength error", async () => {
      const { getByTestId } = renderWithFormik(StreetInput);
      fireEvent.change(getByTestId("street"), {
        target: { value: "street".repeat(7) },
      });
      fireEvent.blur(getByTestId("street"));

      await waitFor(() =>
        expect(getByTestId("streetError")).toHaveTextContent(
          "Must be 40 characters or less"
        )
      );
    });
  });
});

describe("<CityInput />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <Formik validationSchema={addressFormValidator}>
        <CityInput />
      </Formik>
    );
    expect(tree).toMatchSnapshot();
  });

  describe("errors", () => {
    it("should show required error", async () => {
      const { getByTestId } = renderWithFormik(CityInput);

      fireEvent.focus(getByTestId("city"));
      fireEvent.blur(getByTestId("city"));

      await waitFor(() =>
        expect(getByTestId("cityError")).toHaveTextContent("Required")
      );
    });

    it("shows special character error", async () => {
      const { getByTestId } = renderWithFormik(CityInput);

      fireEvent.change(getByTestId("city"), { target: { value: "city@" } });
      fireEvent.blur(getByTestId("city"));

      await waitFor(() =>
        expect(getByTestId("cityError")).toHaveTextContent(
          "Numbers and special characters are not allowed"
        )
      );
    });

    it("shows maxlength error", async () => {
      const { getByTestId } = renderWithFormik(CityInput);
      fireEvent.change(getByTestId("city"), {
        target: { value: "city".repeat(13) },
      });
      fireEvent.blur(getByTestId("city"));

      await waitFor(() =>
        expect(getByTestId("cityError")).toHaveTextContent(
          "Must be 50 characters or less"
        )
      );
    });
  });
});

describe("<StateInput />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <Formik validationSchema={addressFormValidator}>
        <StateInput />
      </Formik>
    );
    expect(tree).toMatchSnapshot();
  });

  describe("errors", () => {
    it("should show required error", async () => {
      const { getByTestId } = renderWithFormik(StateInput);

      fireEvent.focus(getByTestId("stateName"));
      fireEvent.blur(getByTestId("stateName"));

      await waitFor(() =>
        expect(getByTestId("stateNameError")).toHaveTextContent("Required")
      );
    });

    it("shows special character error", async () => {
      const { getByTestId } = renderWithFormik(StateInput);

      fireEvent.change(getByTestId("stateName"), {
        target: { value: "stateName@" },
      });
      fireEvent.blur(getByTestId("stateName"));

      await waitFor(() =>
        expect(getByTestId("stateNameError")).toHaveTextContent(
          "Invalid State name"
        )
      );
    });

    it("shows maxlength error", async () => {
      const { getByTestId } = renderWithFormik(StateInput);
      fireEvent.change(getByTestId("stateName"), {
        target: { value: "stateName".repeat(2) },
      });
      fireEvent.blur(getByTestId("stateName"));

      await waitFor(() =>
        expect(getByTestId("stateNameError")).toHaveTextContent(
          "Must be 15 characters or less"
        )
      );
    });
  });
});

describe("<ZipInput />", () => {
  it("renders correctly", () => {
    const tree = renderer.create(
      <Formik validationSchema={addressFormValidator}>
        <ZipInput />
      </Formik>
    );
    expect(tree).toMatchSnapshot();
  });

  describe("errors", () => {
    it("shows must contain 5 digits error", async () => {
      const { getByTestId } = renderWithFormik(ZipInput);
      fireEvent.change(getByTestId("zip"), { target: { value: "123456" } });
      fireEvent.blur(getByTestId("zip"));

      await waitFor(() =>
        expect(getByTestId("zipError")).toHaveTextContent(
          "Must contain 5 digits"
        )
      );
    });

    it("should show required error", async () => {
      const { getByTestId } = renderWithFormik(ZipInput);
      fireEvent.focus(getByTestId("zip"));
      fireEvent.blur(getByTestId("zip"));

      await waitFor(() =>
        expect(getByTestId("zipError")).toHaveTextContent("Required")
      );
    });
  });
});
