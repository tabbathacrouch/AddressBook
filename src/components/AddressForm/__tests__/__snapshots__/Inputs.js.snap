// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityInput /> renders correctly 1`] = `
<div
  className="city"
>
  <input
    className="field"
    data-testid="city"
    name="city"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="City"
    type="text"
  />
</div>
`;

exports[`<NameInput /> renders correctly 1`] = `
<div>
  <input
    className="field"
    data-testid="name"
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Name"
    type="text"
  />
</div>
`;

exports[`<StateInput /> renders correctly 1`] = `
<div
  className="stateName"
>
  <input
    className="field"
    data-testid="stateName"
    name="stateName"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="State"
    type="text"
  />
</div>
`;

exports[`<StreetInput /> renders correctly 1`] = `
<div>
  <input
    className="field"
    data-testid="street"
    name="street"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Street"
    type="text"
  />
</div>
`;

exports[`<ZipInput /> renders correctly 1`] = `
<div
  className="zip"
>
  <input
    className="field"
    data-testid="zip"
    name="zip"
    onBlur={[Function]}
    onChange={[Function]}
    placeholder="Zip"
    type="text"
  />
</div>
`;
